File: src/robot_life/runtime/life_state.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from robot_life.common.payload_contracts import DetectionPayloadAccessor, ScenePayloadAccessor
from robot_life.common.schemas import EventPriority

SOCIAL_BEHAVIORS = {
    "perform_greeting",
    "greeting_visual_only",
    "perform_attention",
    "attention_minimal",
    "perform_gesture_response",
    "gesture_visual_only",
}
NOTICE_EVENTS = {"familiar_face_detected", "stranger_face_detected", "motion_detected"}
MUTUAL_EVENTS = {"gaze_hold_start_detected", "gaze_sustained_detected"}
ENGAGEMENT_EVENTS = {"wave_detected", "gesture_detected"}
SAFETY_SCENES = {"safety_alert_scene"}
ATTENTION_SCENES = {"attention_scene", "stranger_attention_scene"}
ENGAGEMENT_SCENES = {"greeting_scene", "gesture_bond_scene"}
NOTICED_SCENES = {"ambient_tracking_scene", "attention_scene", "stranger_attention_scene"}


@dataclass
class LifeStateSnapshot:
    latest_scene: Any | None
    latest_scene_payload: dict[str, Any]
    latest_interaction_state: str
    latest_scene_path: str
    latest_target_id: str | None
    latest_engagement_score: Any
    stable_event_types: set[str]
    social_execution: Any | None
    noticed_target_id: str | None
    mutual_target_id: str | None
    engagement_target_id: str | None
    has_p0_event: bool
    has_safety_scene: bool
    has_attention_lost: bool
    has_engagement_scene: bool
    has_mutual_attention_signal: bool
    has_notice_signal: bool


def latest_target_id_from_events(events: list[Any], *, preferred_event_types: set[str]) -> str | None:
    for event in reversed(events):
        if getattr(event, "event_type", None) not in preferred_event_types:
            continue
        payload = getattr(event, "payload", None)
        accessor = DetectionPayloadAccessor(payload if isinstance(payload, dict) else {})
        if accessor.target_id:
            return accessor.target_id
    return None


def build_life_state_snapshot(result: Any) -> LifeStateSnapshot:
    latest_scene = result.scene_candidates[-1] if result.scene_candidates else None
    latest_scene_payload = latest_scene.payload if latest_scene is not None and isinstance(latest_scene.payload, dict) else {}
    latest_scene_accessor = ScenePayloadAccessor(latest_scene_payload)
    latest_interaction_state = latest_scene_accessor.interaction_state or ""
    latest_scene_path = latest_scene_accessor.scene_path or ""
    latest_target_id = getattr(latest_scene, "target_id", None) if latest_scene is not None else None
    latest_engagement_score = latest_scene_accessor.engagement_score
    stable_event_types = {event.event_type for event in result.stable_events}
    social_execution = next((execution for execution in result.execution_results if execution.behavior_id in SOCIAL_BEHAVIORS), None)
    noticed_target_id = latest_target_id_from_events(result.stable_events, preferred_event_types=NOTICE_EVENTS) or latest_target_id
    mutual_target_id = latest_target_id_from_events(result.stable_events, preferred_event_types=MUTUAL_EVENTS) or latest_target_id
    engagement_target_id = latest_target_id_from_events(result.stable_events, preferred_event_types=ENGAGEMENT_EVENTS) or latest_target_id
    has_p0_event = any(event.priority == EventPriority.P0 for event in result.stable_events)
    has_safety_scene = any(scene.scene_type in SAFETY_SCENES for scene in result.scene_candidates)
    has_attention_lost = "attention_lost_detected" in stable_event_types or "gaze_hold_end_detected" in stable_event_types
    has_engagement_scene = any(scene.scene_type in ENGAGEMENT_SCENES for scene in result.scene_candidates) or "wave_detected" in stable_event_types
    has_mutual_attention_signal = (
        latest_interaction_state in {"engaging", "mutual_attention"}
        or any(scene.scene_type in ATTENTION_SCENES for scene in result.scene_candidates)
        or bool(MUTUAL_EVENTS & stable_event_types)
    )
    has_notice_signal = (
        any(scene.scene_type in NOTICED_SCENES for scene in result.scene_candidates)
        or bool(NOTICE_EVENTS & stable_event_types)
    )
    return LifeStateSnapshot(
        latest_scene=latest_scene,
        latest_scene_payload=latest_scene_payload,
        latest_interaction_state=latest_interaction_state,
        latest_scene_path=latest_scene_path,
        latest_target_id=latest_target_id,
        latest_engagement_score=latest_engagement_score,
        stable_event_types=stable_event_types,
        social_execution=social_execution,
        noticed_target_id=noticed_target_id,
        mutual_target_id=mutual_target_id,
        engagement_target_id=engagement_target_id,
        has_p0_event=has_p0_event,
        has_safety_scene=has_safety_scene,
        has_attention_lost=has_attention_lost,
        has_engagement_scene=has_engagement_scene,
        has_mutual_attention_signal=has_mutual_attention_signal,
        has_notice_signal=has_notice_signal,
    )

File: Robot_Life_CPP/src/runtime/life_state.cpp
#include "robot_life_cpp/runtime/life_state.hpp"

#include <algorithm>
#include <set>

namespace robot_life_cpp::runtime {

namespace {

std::optional<std::string> latest_target_id_from_events(
    const std::vector<common::StableEvent>& events,
    const std::unordered_set<std::string>& preferred_event_types) {
  for (auto it = events.rbegin(); it != events.rend(); ++it) {
    if (!preferred_event_types.contains(it->event_type)) {
      continue;
    }
    auto target_it = it->payload.find("target_id");
    if (target_it != it->payload.end() && !target_it->second.empty()) {
      return target_it->second;
    }
  }
  return std::nullopt;
}

std::optional<double> parse_optional_double(const common::Payload& payload, const std::string& key) {
  const auto it = payload.find(key);
  if (it == payload.end() || it->second.empty()) {
    return std::nullopt;
  }
  try {
    return std::stod(it->second);
  } catch (...) {
    return std::nullopt;
  }
}

}  // namespace

LifeStateSnapshot build_life_state_snapshot(
    const std::vector<common::StableEvent>& stable_events,
    const std::vector<common::SceneCandidate>& scene_candidates,
    const std::vector<common::ExecutionResult>& execution_results,
    const common::SceneTaxonomyRules& taxonomy) {
  LifeStateSnapshot snapshot{};

  if (!scene_candidates.empty()) {
    snapshot.latest_scene = scene_candidates.back();
    snapshot.latest_scene_payload = scene_candidates.back().payload;
    auto state_it = snapshot.latest_scene_payload.find("interaction_state");
    if (state_it != snapshot.latest_scene_payload.end()) {
      snapshot.latest_interaction_state = state_it->second;
    }
    auto path_it = snapshot.latest_scene_payload.find("scene_path");
    if (path_it != snapshot.latest_scene_payload.end()) {
      snapshot.latest_scene_path = path_it->second;
    }
    snapshot.latest_target_id = scene_candidates.back().target_id;
    snapshot.latest_engagement_score = parse_optional_double(snapshot.latest_scene_payload, "engagement_score");
  }

  std::set<std::string> event_set{};
  snapshot.has_p0_event = false;
  for (const auto& event : stable_events) {
    event_set.insert(event.event_type);
    if (event.priority == common::EventPriority::P0) {
      snapshot.has_p0_event = true;
    }
  }
  snapshot.stable_event_types.assign(event_set.begin(), event_set.end());

  for (const auto& execution : execution_results) {
    if (taxonomy.social_behaviors.contains(execution.behavior_id)) {
      snapshot.social_execution = execution;
      break;
    }
  }

  snapshot.noticed_target_id = latest_target_id_from_events(stable_events, taxonomy.notice_events);
  if (!snapshot.noticed_target_id.has_value()) {
    snapshot.noticed_target_id = snapshot.latest_target_id;
  }

  snapshot.mutual_target_id = latest_target_id_from_events(stable_events, taxonomy.mutual_events);
  if (!snapshot.mutual_target_id.has_value()) {
    snapshot.mutual_target_id = snapshot.latest_target_id;
  }

  snapshot.engagement_target_id = latest_target_id_from_events(stable_events, taxonomy.engagement_events);
  if (!snapshot.engagement_target_id.has_value()) {
    snapshot.engagement_target_id = snapshot.latest_target_id;
  }

  snapshot.has_safety_scene = std::any_of(
      scene_candidates.begin(),
      scene_candidates.end(),
      [&](const common::SceneCandidate& scene) { return taxonomy.safety_scenes.contains(scene.scene_type); });

  const bool has_attention_lost_event =
      event_set.contains("attention_lost_detected") || event_set.contains("gaze_hold_end_detected");
  snapshot.has_attention_lost = has_attention_lost_event;

  snapshot.has_engagement_scene =
      std::any_of(
          scene_candidates.begin(),
          scene_candidates.end(),
          [&](const common::SceneCandidate& scene) { return taxonomy.engagement_scenes.contains(scene.scene_type); }) ||
      event_set.contains("wave_detected");

  const bool has_attention_scene = std::any_of(
      scene_candidates.begin(),
      scene_candidates.end(),
      [&](const common::SceneCandidate& scene) { return taxonomy.attention_scenes.contains(scene.scene_type); });
  snapshot.has_mutual_attention_signal =
      snapshot.latest_interaction_state == "engaging" ||
      snapshot.latest_interaction_state == "mutual_attention" ||
      has_attention_scene ||
      event_set.contains("gaze_hold_start_detected") ||
      event_set.contains("gaze_sustained_detected");

  const bool has_noticed_scene = std::any_of(
      scene_candidates.begin(),
      scene_candidates.end(),
      [&](const common::SceneCandidate& scene) { return taxonomy.noticed_scenes.contains(scene.scene_type); });
  snapshot.has_notice_signal =
      has_noticed_scene ||
      event_set.contains("familiar_face_detected") ||
      event_set.contains("stranger_face_detected") ||
      event_set.contains("motion_detected");

  return snapshot;
}

}  // namespace robot_life_cpp::runtime
